### PR TITLE
Make init method on IPushMessages async

### PR DESCRIPTION
**Pull request by tmasternak**
This is needed to enable queue purging in sql transport. [see](https://github.com/Particular/NServiceBus.SqlServer/pull/108/files?diff=unified#diff-7699ae3c475c6f3261c50c3b69adec1eR35)

### CODE REVIEW

**Comment by tmasternak** [comment]
@danielmarbach could you have look?

File: src/NServiceBus.Core/Transports/TransportReceiver.cs [hunk 1]
```diff
@@ -38,7 +38,7 @@ class TransportReceiver
 
             Logger.DebugFormat("Pipeline {0} is starting receiver for queue {1}.", Id, pushSettings.InputQueue);
 
-            receiver.Init(InvokePipeline, pushSettings);
+            await receiver.Init(InvokePipeline, pushSettings);
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Transports/TransportReceiver.cs]: configure await?

**Comment by danielmarbach** [comment]
Another design option would be to purge when we start? Thoughts?

**Comment by tmasternak** [comment]
I think that Init is better. Here are my thoughts on alternatives:
* we need Start to be async - this is pretty much equivalent to Init being async but in my opinion Init is a more suitable place in terms of responsibility.
* we keep Start void and we handle purging in main processing task - this seems to complicate things a bit more than async Init especially when it comes to handling exceptions during purging. Should we throw exception (probably not)? Show we keep trying and notify curcuitbreaker?

**Comment by tmasternak** [comment]
@danielmarbach do you think the current approach is fine in context of previous comment?

**Comment by danielmarbach** [comment]
Definitely. Sorry free days after NSBCon

On Tuesday, December 8, 2015, Tomasz Masternak <notifications@github.com>
wrote:

> @danielmarbach <https://github.com/danielmarbach> do you think the
> current approach is fine in context of previous comment?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3158#issuecomment-162885648>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by yvesgoeleven** [comment]
@tmasternak While you're at it, can you pass in CriticalError callback in Init as well, so that all callbacks are all initialized in the same place

**Comment by andreasohlund** [comment]
@yvesgoeleven this pull was merged 2 days ago, raise a new pull for that proposal?

**Comment by yvesgoeleven** [comment]
@andreasohlund sorry, I was wading my way through 600 unread github notifications, did not pay attention to that

**Comment by yvesgoeleven** [comment]
@andreasohlund I've quickly implemented it, when build of my branch is done, I'll send the PR

**Comment by SimonCropp** [comment]
@yvesgoeleven why do u need CriticalError  in the init? it can be accessed via the container

**Comment by yvesgoeleven** [comment]
@SimonCropp transports don't have access to the container

**Comment by danielmarbach** [comment]
We want to reduce the container usages and favour explicit interfaces between core and the transport.

**Pull request merged**